Sources_v3/Controllers/ListDatabaseObserver.swift

```diff
@@ -186,6 +199,50 @@ class ListChangeAggregator<DTO: NSManagedObject, Item>: NSObject, NSFetchedResul
     func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
         onChange?(currentChanges)
     }
+    
+    // MARK: - Private
+    
+    private func subscribeToNotifications() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(willSaveContext),
+            name: .NSManagedObjectContextWillSave,
+            object: nil
+        )
+    }
+    
+    private func unsubscribeFromNotifications() {
+        NotificationCenter.default.removeObserver(self)
+    }
+    
+    @objc private func willSaveContext(_ notification: Notification) {
+        guard let context = notification.object as? NSManagedObjectContext else { return }
+    
+        // Get `DTO` ids the FRC is observing
+        let fetchedDTOs = frc.fetchedObjects ?? []
+        
+        // Get `DTO` ids that are gonna be deleted
+        let deletedDTOs = context
+            .deletedObjects
+            .compactMap { $0 as? DTO }
+        
+        // Get `DTO` ids that that no longer match the fetch request predicate
+        let missmatchingDTOs = context
+            .updatedObjects
+            .compactMap { $0 as? DTO }
+            .filter { frc.fetchRequest.predicate?.evaluate(with: $0) == false }
+        
+        // Perform `DTO -> Item` transformation in advance
+        deletedItems.removeAll()
+        (deletedDTOs + missmatchingDTOs).forEach { dtoToDelete in
+            guard let trackedDTO = fetchedDTOs.first(where: { $0.objectID == dtoToDelete.objectID }) else { return }
+            
+            // Create item from DTO in advance before the actual deletion happen.
+            // This is required because after the deletion, entity relations become invalid
+            // and lossless `DTO -> Item` converion become impossible
+            deletedItems[trackedDTO.objectID] = itemCreator(trackedDTO)
+        }
+    }
```

Comment:
The problem I'm solving here is that the entity relationships become invalid after the `DTO -> Item` transformation is not possible.

I don't like neither `No action` delete rule nor `N-step-database-flush` so come up with performing `DTO -> Item` transformation **before** the actual deletion happen.